### ISSUE TITLE
epic(trusty-search): default Apple-Silicon Python/MPS embedder

### ISSUE THREAD

**Issue by bobmatnyc** [bobmatnyc/trusty-tools#3524]
## Decision and outcome

Ship the validated Python `sentence-transformers` / PyTorch-MPS sidecar as the default embedder on Apple Silicon, managed by `uv` and speaking the existing newline-JSON-RPC-over-stdio supervisor protocol.

Defaults:

- Apple Silicon: sidecar enabled by default with fp32;
- fp16: explicit throughput option;
- unsupported or unhealthy environments: clear fallback and diagnostics.

## Scope

- reproducible environment/bootstrap and model lifecycle;
- packaging, discovery, health checks, and supervisor integration;
- safe fallback to the supported Rust path;
- configuration and status surfaces;
- performance, correctness, cold-start, and failure benchmarks.

## Done when

A clean Apple-Silicon install selects the sidecar without manual setup, produces compatible embeddings, survives restart/failure scenarios, and meets the recorded benchmark threshold without regressing fallback platforms.

The former delivery tickets #3530, #3553, #3569, and #4106 are consolidated here.

**Comment by bobmatnyc** [comment]
## Progress update

### Slice 1 (spike/proof) → **DONE**
Validated on M4 Max through the real supervisor:
- End-to-end supervisor (reindex incl. HNSW): **561 emb/s** — beats the 457 emb/s post-CPU-fix Rust baseline target and is ~6.3× the original 88.5 INT8-CoreML default.
- Direct wire, single-flight: fp32/bs512 = **1094 emb/s**, fp16/bs512 = **1336 emb/s**; pure in-process fp16 = **1758** (≈ the 1705 PyTorch-MPS ceiling).
- Correctness vs `REFERENCE_VECTORS`, measured on MPS: fp32 mean & min cosine = **1.0000000**; fp16 ≥ **0.9999990**. All 384-dim, unit-norm.
- Memory: peak sidecar RSS **< 800 MB**, bounded over 24k sustained embeds — no #24-style runaway.
- Zero trusty-search code changes; production daemon untouched.

### Slices 2–4 (production Python module + Rust launcher/wiring + uv bootstrap) → **IN ACTIVE DEVELOPMENT**
Being built as one integrated **default-OFF** PR (a build is in flight; PR link to be added when it opens).

### Slices 5–7 → **pending**
- 5: fallback-to-Rust + doctor polish + `/health` real-provider readback.
- 6: packaging/install + docs + the default-on Apple-Silicon switch.
- 7: benchmark + #24 memory-safety soak (ties to #24).

### Observability note
The `/health` **real-provider readback** — the specific piece of #3493 P1 that applies to this sidecar — lands in **slice 5**. The broader observability cleanup remains tracked on umbrella **#3493**.

🤖🤖🤖 Generated with trusty-mpm — https://github.com/bobmatnyc/trusty-tools

**Comment by bobmatnyc** [comment]
## Linux/CUDA validation finding: sidecar-crash recovery is only half-wired (slice 5 input)

During the Linux/CUDA validation pass for this epic's sidecar path, we exercised
a sidecar death mid-run and found the recovery story is incomplete:

- **Good:** the `EmbedderSupervisor` correctly detects the dead Python sidecar
  process and **respawns** it — the supervisor/restart-backoff machinery works
  as designed.
- **Gap:** the deferred-embed pass that was **in flight at the moment of the
  crash fails outright, with no auto-retry**. The semantic (vector) lane for
  that reindex/index-file operation is left in a terminal `failed` state — it
  does **not** get re-enqueued against the freshly-respawned sidecar. The only
  way to recover is a **manual** full reindex.
- Also confirmed: on sidecar death, there is currently no fallback to the
  Rust `ort` embedder path for the in-flight work — the deferred-embed task
  just fails; it doesn't fail over to the alternate backend the way the
  `python` mode's own bootstrap-failure path does (`build_embedder`'s
  `TRUSTY_EMBEDDER=python` arm, which falls back to
  `build_ort_stdio_sidecar()` on bootstrap/launcher-discovery failure —
  that's a *startup*-time fallback only, not a *mid-flight-crash* fallback).

## Ask for slice 5 (fallback/health)

Please add, as part of slice 5's fallback/health work:

- **Auto-retry / re-enqueue** the failed deferred-embed pass after a sidecar
  respawn, so a transient sidecar crash doesn't leave an index's semantic
  lane permanently `failed` until an operator notices and manually triggers a
  reindex. A bounded retry (a handful of attempts with backoff) against the
  now-healthy respawned sidecar should be enough to self-heal the common
  transient-crash case.
- Surface the failed/retrying state distinctly (vs. a genuine embedding
  error) so operators can tell "transient sidecar crash, retrying" apart from
  "this file/content is actually broken."

Filed as part of the broader CUDA validation sweep — see the umbrella #3493
for the full pass/fail summary (native ort CUDA + this sidecar both validated
functionally on a T4; this recovery gap was the one behavioral issue found).

**Comment by bobmatnyc** [comment]
Slices 2–4 merged via #3529 (squash SHA `4bdca8f895e5753a72c54d76d985364c7576491a`), rebased onto the 0.36.1 release main.

## Code Review Summary

Code-critic reviewed across two rounds; all WARN findings fixed:
- **SIGTERM handler**: real-signal readiness test in place
- **Orphan fast-follows**: 
  - `TOKENIZERS_PARALLELISM=1` environment hardening
  - Progress-aware shutdown watchdog
- **Performance/robustness**:
  - `.ready` lightweight per-respawn recheck
  - `TRUSTY_EMBEDDER_PY_IDLE_SHUTDOWN_SECS=1800` idle-shutdown default
- **Documentation**: CLAUDE.md docs updated

## Status

Ships **DEFAULT-OFF**. Full Python sidecar path gated behind explicit `TRUSTY_EMBEDDER=python` env or future UI switch.

## Remaining Work (slices 5–7)

- **Slice 5** (fallback/doctor polish): `/health` real-provider readback + deferred-embed auto-retry
- **Slice 6** (packaging/install): signing + default-on-Apple-Silicon switch
- **Slice 7** (bench + #24 soak): throughput bench + memory-safety sustained load

**Comment by bobmatnyc** [comment]
## Repo-owner sign-off: `SupervisorHandle::has_given_up()` accepted as a scoped exception

Slice 5 (PR #3560) adds `SupervisorHandle::has_given_up()` to `crates/trusty-common/src/embedder_client/supervisor.rs` (+62 lines). This epic's acceptance criteria include the phrase **"zero supervisor code changes"**, so the addition was escalated for an explicit decision rather than assumed acceptable.

**Decision (Bob, 2026-07-21): approved as a scoped exception.**

Rationale of record:
- The constraint is listed under the **Protocol conformance** acceptance bullet (`id-echo / out-of-order / empty-batch / EOF-shutdown; zero supervisor code changes`), so it is read as scoped to protocol handling — not an absolute prohibition on the file.
- The change is **additive and observational only**: one new method plus a second one-way `watch` channel mirroring the `shutdown_tx` pattern already present in the same struct.
- `should_give_up`'s decision logic, restart counting, and backoff behavior are **byte-identical**. The supervisor's own respawn behavior does not change.

Why it was necessary: adversarial review found the original fallback latch counted *consecutive failed embed requests* while claiming to mirror the supervisor's give-up ceiling (*consecutive respawn attempts*). Because `LazyEmbedderHandle::embed_via` hands out the current client without waiting for a respawn, concurrent in-flight requests during a single crash episode each failed independently — so ~6 simultaneous failures could permanently latch to the slower Rust ort path on the **first** crash, silently and irreversibly downgrading the Apple-Silicon default path for the daemon's lifetime. Reading the supervisor's real give-up signal is the correct fix; the request-level proxy was rejected as incorrect under concurrency.

Slices 6 and 7 should treat the constraint as protocol-scoped on the same basis.

🤖🤖🤖 Generated with trusty-mpm — https://github.com/bobmatnyc/trusty-tools

**Comment by bobmatnyc** [comment]
## Slice 5 Shipped

**Slice 5 merged** as squash commit `909dd4cb` via PR #3560 ("feat: Python/MPS sidecar runtime fallback + doctor check + real /health readback"). All 23 CI checks green at merge.

**What shipped:** the `FallbackEmbedderAdapter` one-way latch to the Rust ort path; the `trusty-search doctor` `python_embedder` check with `--fix`; and `/health` real-provider readback (wire-level device echoed by the Python sidecar and captured by the stdio client).

**A production bug was found and fixed along the way** — worth calling out explicitly as the highest-value outcome. In `supervision_loop` (`crates/trusty-common/src/embedder_client/supervisor.rs`), the top-of-loop match treated an empty `child_slot` as proof of cooperative shutdown and returned. But the `Unhealthy` wedge-kill branch also empties that slot before respawning, so a *failed* respawn caused the next iteration to silently return — never incrementing `consecutive_failures`, never reaching `should_give_up`, never firing the give-up signal. Real-world impact: a daemon left with a dead sidecar, no supervision, and fallback that never triggers. Fixed via a new `RestartTrigger::RespawnFailed` variant.

**Design note for slices 6-7:** the fallback latch trips on the supervisor's real give-up signal (`supervisor_gave_up()`), *not* an independently-counted request-failure proxy. Adversarial review found the original request-counting design could permanently latch to the slow Rust path on the first crash under concurrent load. Do not reintroduce a request-level proxy.

**Slice 5 / slice 6 reconciliation:** slice 6 PR 2/5 (#3563) shipped overlapping `/health` backend reporting while slice 5 was in review. These were reconciled rather than duplicated — `SwitchableEmbedder::resolved_provider_label()` now forwards to the active backend so slice 5's live wire-readback is not shadowed by slice 6's `ActiveBackend` snapshot prediction. Flag for whoever owns slices 6-7 that the two lanes have overlapping surface and coordination is advisable before slice 7.

**Known unrelated issue:** #3564 (unbounded `.gitignore` walk-up escaping tempdir) was filed from this work; it is pre-existing and independent.

🤖🤖🤖 Generated with trusty-mpm — https://github.com/bobmatnyc/trusty-tools

**Comment by bobmatnyc** [comment]
## Lane boundary for slices 6-7 (embedder fallback surface)

Proposed boundary after slice 5 landed and slice 6 PR-4 (#3584) was found to have branched pre-slice-5. Stated so whichever session picks up slices 6-7 has it in writing.

**Single source of truth.** The `EmbedderSupervisor`'s `should_give_up` decision (`crates/trusty-common/src/embedder_client/supervisor.rs`) is the one authority for "the sidecar is genuinely dead." Expose it through exactly **one** readback — `SupervisorHandle::has_given_up()` / `Embedder::supervisor_gave_up()`. Do not add parallel signals for the same event.

**Ownership by path:**
- `crates/trusty-common/src/embedder_client/supervisor.rs` — owns the decision and the single give-up readback. **Highest-contention file in the epic**: three lanes have now landed or attempted changes here (#3560, #3575 ETXTBSY retry, the slice-6 train). Any future PR touching it should be diffed against current `main`, not a stale base, before review.
- `crates/trusty-search/src/commands/start/embedder_fallback.rs` — runtime fallback for the **EagerPython** arm only. Do not extend it to wrap the graceful arm.
- `crates/trusty-search/src/commands/start/swap_back_watchdog.rs` + `graceful_bootstrap.rs` — runtime fallback for the **GracefulPython** arm only, via `SwitchableEmbedder::swap_to`.
- If the two arms are ever unified, the seam is: the `has_given_up` decision stays shared, but exactly one consumer decides whether to latch a wrapper (slice-5 style) or swap the switchable (slice-6 style) — never both wired independently against the same instance.

**Binding design constraint (from slice 5's adversarial review):** fallback must trip on the supervisor's real give-up signal, never an independently-counted request-failure proxy. The proxy design could permanently latch to the slow ort path on the first crash under concurrent multi-flight load. Do not reintroduce it.

**Slice 7 (benchmark + #24 memory soak)** needs one fallback path to soak, not two. The realistic target is the **GracefulPython** arm, since `TRUSTY_PY_DEFAULT` is the intended Apple-Silicon default. It must not duplicate: a third give-up signal (reuse `has_given_up()`); its own death-detection heuristic; or its own "build a fresh ort backend" helper (reuse `build_ort_stdio_sidecar()`, already shared). The soak should exercise repeated ort→python→ort→python cycles under sustained MPS reindex, to catch a leaked python child (`PythonAdapterTeardown::teardown()` not awaited) or an orphaned ort backend from a failed `build_ort()`.

🤖🤖🤖 Generated with trusty-mpm — https://github.com/bobmatnyc/trusty-tools

**Comment by bobmatnyc** [comment]
**Lane notice — slice 7 starting (session tm-tcode-01).**

Picking up item 7 (benchmark + #24 re-check: throughput bench + memory-safety soak under sustained MPS reindex). Flagging boundaries so this doesn't repeat the slice-5/slice-6 `/health` overlap:

- **I am NOT touching slice 6's remaining scope** — packaging/install, launcher signing, cargo-install wiring, docs, or the default-on-Apple-Silicon flip. PR #3624 (`feat-py-default-flip-3524-pr5b`) owns the default flip; I will not modify it or its files.
- **Slice 7 targets the `GracefulPython` arm** (`SwitchableEmbedder` + `swap_back_watchdog`), not the eager arm, and adds bench/soak coverage rather than changing supervisor behavior.
- **No new give-up signal will be introduced.** Verified on current main: `terminated_signal()` (`supervisor.rs:202`) and `has_given_up()` (`supervisor.rs:355`) are complementary views of one decision, both written from the single `should_give_up` trip site (`supervisor.rs:779,785`), both monotonic. The swap-back watchdog trips on `is_confirmed_terminated()`; slice 5's `FallbackEmbedderAdapter` still trips on `supervisor_gave_up()`. Slice 7 reuses these as-is — the binding constraint from slice 5 holds and stays held.

If anyone else is already on slice 7, say so here and I'll stand down.

🤖🤖🤖 Generated with trusty-mpm — https://github.com/bobmatnyc/trusty-tools

**Comment by bobmatnyc** [comment]
Triage consolidation (2026-07-28): this issue is now the canonical, outcome-oriented tracker for its former implementation leaves. Scope and acceptance criteria were retained here; leaf tickets are being closed as `not planned` to reduce duplicate queue surface, not to claim the work is implemented.

**Comment by bobmatnyc** [comment]
Selection gap, verified 2026-08-11 against `trusty-search-v0.45.0` — distinct from the reporting/silence half, which is already fixed and shipping (see #4125, PR #4780/#4797, trusty-search 0.42.1+).

`trusty-embedderd-py`'s launcher ships in NO release archive, NO installer, and no macOS codesigning path. `.github/workflows/release.yml`'s trusty-search `BINARIES` reads:

```
BINARIES="trusty-search trusty-embedderd"
```

No `trusty-embedderd-py`. `gh release view trusty-search-v0.45.0 --json assets` lists only:

```
trusty-search-0.45.0-aarch64-apple-darwin.tar.gz(.sha256)
trusty-search-0.45.0-aarch64-linux-al2023.tar.gz(.sha256)
trusty-search-0.45.0-aarch64-unknown-linux-gnu.tar.gz(.sha256)
trusty-search-0.45.0-x86_64-linux-al2023.tar.gz(.sha256)
trusty-search-0.45.0-x86_64-unknown-linux-gnu.tar.gz(.sha256)
```

No python-launcher artifact for any target. Repo-wide, `embedderd-py`/`embedderd_py` appears nowhere in `.github/workflows/`, `Makefile`, or `crates/trusty-installer/src`.

Consequence: an Apple Silicon installer/release-archive user gets ORT instead of MPS/torch. That is now correctly REPORTED as degraded (shipped 0.42.1, PR #4780/#4797) — but the user has no way to obtain the intended backend, because the binary that would provide it is not distributed to them.

This is the SELECTION half, not the reporting half. Fixing the reporting did not and could not fix this — the two are independent defects with independent fixes. Someone reading #4125's resolution should not conclude this is handled.

The finding dates to 2026-07-27 (originally #4106, closed as consolidated into this epic) and remains true against today's v0.45.0 — it has survived several trusty-search releases since as a folded-in epic mention rather than a tracked, actionable item.

🤖🤖🤖 Generated with trusty-mpm — https://github.com/bobmatnyc/trusty-tools

**Comment by bobmatnyc** [comment]
Owner ruling, 2026-08-30: the Python/MPS-as-default-embedder direction is abandoned; ORT + fp16 is the shipped path. The four consolidated delivery tickets (#3530, #3553, #3569, #4106) are all closed, but the epic's own "Done when" bar — the Python sentence-transformers/PyTorch-MPS sidecar becomes the DEFAULT embedder on Apple Silicon — was never met and is not being pursued: `crates/trusty-search/src/service/embedder_supervisor/switchable.rs` documents the Rust ONNX-runtime stdio sidecar (`trusty-embedderd`) as "the default," not the Python/MPS path. Closed per owner ruling, 2026-08-30 enhancement pass.
